hkmc2/shared/src/test/mlscript-compile/Iter.mls

```diff
@@ -0,0 +1,95 @@
+import "./Predef.mls"
+
+open Predef
+
+// * Implements the `Iterable` protocol.
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol
+class Iterable(mk) with
+  set this.(globalThis.Symbol.iterator) = mk
+
+// * Implements the `Iterator` protocol.
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol
+class Iterator(val next)
+
+module Result with
+  class Next(val value) with
+    val done = false
+  object Done with
+    val done = true
+
+module Iter with ...
+
+fun derive(iterable, makeNext) = Iterable of () =>
```

Comment:
The name is quite confusing. Why not
```suggestion
fun adaptIterator(iterable, makeNext) = Iterable of () =>
```

Reply:
Also why does it require all callers to wrap things in an obfuscating `() =>`?

Reply:
This function takes the old iterator and produces a function that is going to be used as the `next` method on the new `Iterator`.

Reply:
(I didn't use the Commit suggestion feature to make the changes because the references below also needs to be updated.)